### PR TITLE
Module4 homework

### PR DESCRIPTION
**Pull request by markrenaud**
Module 4 homework ready for review.

### CODE REVIEW

File: Module-4/Homework.playground/Contents.swift [hunk 1]
```diff
@@ -1,2 +1,220 @@
...
+ - Call the function `commonElementsSet` passing the arrays `numsBy2`, `numsBy4`, and print out the results.
+ */
+let nums = Array(0...100)
+let numsWithNil = [79, nil, 80, nil, 90, nil, 100, 72]
+let numsBy2 = Array(stride(from: 2, through: 100, by: 2))
+let numsBy4 = Array(stride(from: 2, through: 100, by: 4))
+
+func evenNumbersArray(_ array: [Int]) -> [Int] {
+    array.filter { $0 % 2 == 0 }
+}
```
Comment (niknovak) [inline Module-4/Homework.playground/Contents.swift:131]: Remember to call it and print as well ;)

File: Module-4/Homework.playground/Contents.swift [hunk 2]
```diff
@@ -1,2 +1,220 @@
...
+
+func sumOfArray(_ array: [Int?]) -> Int {
+    array.reduce(0) { runningTotal, element in
+        runningTotal + (element ?? 0)
+    }
```
Comment (niknovak) [inline Module-4/Homework.playground/Contents.swift:136]: One way to also eliminate nil values from the array is to use a `compactMap`. Then, the `reduce` method can be used with a shorthand syntax.
For example:
```swift
return array.compactMap { $0 }.reduce(0, +)
```

File: Module-4/Homework.playground/Contents.swift [hunk 3]
```diff
@@ -1,2 +1,220 @@
...
+ 
+ Next, extend the protocol `Shape` to add a new method called `calculateVolume() -> Double`.
+ 
+ Finally, create a `struct` called `Sphere` that conforms to `Shape`. `Sphere` should have appropriate stored properties for calculating area and volume.
+ 
+ Create an instance of `Sphere` and print out the area and volume.
+ */
+protocol Shape {
+    func calculateArea() -> Double
+    func calculateVolume() -> Double
```
Comment (niknovak) [inline Module-4/Homework.playground/Contents.swift:171]: You don't actually need to declare the function here when you do it in an extension.

**Review by niknovak — APPROVED** [review]
Well done!

File: Module-4/Homework.playground/Contents.swift (hunk 2 above)
Comment (markrenaud) [reply Module-4/Homework.playground/Contents.swift:136]: I appreciate the suggestion to use compactMap followed by reduce. Initially, I chose the reduce with a nil-coalescing operator to minimise the number of iterations through the array, as the compactMap + reduce combination would iterate twice.

After you comment, I decided to test the performance of each option:

```Swift
final class SummationTests: XCTestCase {

    let optionalSequence: [Int?] = {
        var arr: [Int?] = Array(0...100000)
        // insert some nil values throughout the array
        for idx in stride(from: 0, to: 10, by: (1...5).randomElement()!) {
            arr[idx] = nil
        }
        return arr
    }()

    func testTernaryReduce() {
        measure {
            let sum = optionalSequence.reduce(0) { runningTotal, element in
                runningTotal + (element ?? 0)
            }
        }
    }
    
    func testCompactMapVariant() {
        measure {
            let sum = optionalSequence.compactMap { $0 }.reduce(0, +)
        }
    }

}
```

The test results on my Mac indicated that the reduce with nil-coalescing was faster (0.012s) compared to the compactMap + reduce approach (0.022s). However, the absolute time difference (0.010s) is quite small and may not justify the loss of readability.

I would now choose the compactMap + reduce approach for its clarity and conciseness in this instance.

**Pull request merged**